Feature: Nft governance badger

fabianschu opened the pull request:
closes #230

marvinkruse commented on contracts/governance/Badger.sol:
```diff
@@ -0,0 +1,342 @@
+        address account,
+        uint256 id,
+        uint256 amount
+    ) public onlyOwner isTier(id) {
```
You are already checking `isTier(id)` in `_burn()` (line 339)

marvinkruse commented on contracts/governance/Badger.sol:
```diff
@@ -0,0 +1,342 @@
+        private
+        isTier(tokenId)
+        isValidString(uriId)
+    {
```
What's the idea here behind using `isTier` and `isValidString`?
It looks to me like both just validate that the URI string is non-empty. Wouldn't one be enough?

marvinkruse replied:
Sorry, I was stupid. The idea is to check that the new uriId is valid. Ignore this :D

fabianschu replied:
They're very similar but not exactly the same. `isTier` checks whether the tier that should be updated actually exists (by looking that the `uriId` is not empty, which is a requirement for a tier) and `isValidString` checks the param, or put differently that the new `uriId` is not empty.

marvinkruse reviewed (dismissed):
looks good to me

Zitzak commented on contracts/governance/Badger.sol:
```diff
@@ -0,0 +1,342 @@
+        );
+        require(
+            msg.sender == owner() ||
+                from == _msgSender() ||
```
I couldn't find where `_msgSender()` is coming from or what it does, altho I can guess. Is it just the same as `msg.sender`?

fabianschu replied:
I changed it to use `msgSender()` consistently.

Zitzak replied:
Thanks for the explanation

fabianschu commented:
- removed `mintMultipleToMultiple` since same functionality can be achieved with `mintToMultiple` by passing in the same address multiple times
- added event for tier creation/update

fabianschu commented on contracts/governance/Badger.sol:
```diff
@@ -0,0 +1,342 @@
+        address[] calldata accounts,
+        uint256[] calldata tokenIds,
+        uint256[] calldata amounts
+    ) public onlyOwner isSameLength(accounts, tokenIds, amounts) {
```
If you use modifiers, their code is basically copied over into each function that makes use of the respective modifier, they are "inlined". This increases the contract size, because on the compiled-code-level it copies the same modifier code into different places. If one would replace the modifier with an internal function, this would reduce contract size, but make each function invocation slightly more expensive, since calling an internal function is a little bit more expensive than running inlined code.

In this context a slightly larger contract is not a problem, imo.

[Here's a good article](https://blog.polymath.network/solidity-tips-and-tricks-to-save-gas-and-reduce-bytecode-size-c44580b218e6), check out the section _Function modifiers can be inefficient_

Zitzak reviewed (approved):
lgtm

marvinkruse reviewed (approved):
lgtm!
